File: Ch03/challenge/magic.py
class MagicError(Exception):
    pass


def file_type(file_name):
    with open(file_name, 'rb') as fp:
        data = fp.read(64)

    if data[:4] == b'\x00\x00\x01\x00':
        return 'ico'
    if data[:6] == b'\x47\x49\x46\x38\x37\x61':
        return 'gif'
    if data[:4] == b'\xFF\xD8\xFF\xF0':
        return 'jpg'
    if data[:8] == b'\x89\x50\x4E\x47\x0D\x0A\x1A\x0A':
        return 'png'

    raise MagicError(f'{file_name!r} - unknown file type')


if __name__ == '__main__':
    file_name = 'Ch03/challenge/head.png'
    try:
        typ = file_type(file_name)
        print(f'{file_name} -> {typ}')
    except (FileNotFoundError, MagicError) as err:
        print(f'error: {err}')

File: Ch03/solution/magic.go
package main

import (
	"bytes"
	"fmt"
	"os"
)

func fileType(fileName string) (string, error) {
	file, err := os.Open(fileName)
	if err != nil {
		return "", err
	}
	defer file.Close()

	data := make([]byte, 64)
	if _, err := file.Read(data); err != nil {
		return "", err
	}

	if bytes.Equal(data[:4], []byte{0x00, 0x00, 0x01, 0x00}) {
		return "ico", nil
	}
	if bytes.Equal(data[:6], []byte{0x47, 0x49, 0x46, 0x38, 0x37, 0x61}) {
		return "gif", nil
	}
	if bytes.Equal(data[:4], []byte{0xFF, 0xD8, 0xFF, 0xF0}) {
		return "jpg", nil
	}
	if bytes.Equal(data[:8], []byte{0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A}) {
		return "png", nil
	}

	return "", fmt.Errorf("%q - unknown file type", fileName)
}

func main() {
	fileName := "head.png"
	typ, err := fileType(fileName)
	if err != nil {
		fmt.Println("ERROR:", err)
		return
	}
	fmt.Printf("%s -> %s\n", fileName, typ)
}
